lib/java/com/google/android/material/lists/MaterialListAdapter.java

```diff
@@ -0,0 +1,73 @@
+package com.google.android.material.lists;
+
+import android.view.ViewGroup;
+
+import androidx.annotation.NonNull;
+import androidx.recyclerview.widget.RecyclerView;
+
+import com.google.android.material.lists.listdatatypes.ArrayListData;
+import com.google.android.material.lists.listdatatypes.CustomListData;
+import com.google.android.material.lists.listdatatypes.ListData;
+
+import java.lang.reflect.InvocationTargetException;
+import java.util.List;
+
+/**
+ * An adapter that uses {@link MaterialViewHolder}s and binds data to them using {@link MaterialListBinder}
+ */
+public class MaterialListAdapter<T, E extends MaterialViewHolder> extends RecyclerView.Adapter<E> {
+
+  private CustomListData<T> customListData;
+  private MaterialListBinder<T, E> materialListBinder;
+  private Class<E> viewHolderClass;
+
+  public MaterialListAdapter(MaterialListBinder<T, E> materialListBinder, Class<E> viewHolderClass, T[] data) {
+    this.materialListBinder = materialListBinder;
+    this.viewHolderClass = viewHolderClass;
+    customListData = new ArrayListData<>(data);
+  }
+
+  public MaterialListAdapter(MaterialListBinder<T, E> materialListBinder, Class<E> viewHolderClass, List<T> data) {
+    this.materialListBinder = materialListBinder;
+    this.viewHolderClass = viewHolderClass;
+    customListData = new ListData<>(data);
+  }
+
+  public MaterialListAdapter(MaterialListBinder<T, E> materialListBinder, Class<E> viewHolderClass, CustomListData<T> data) {
+    this.materialListBinder = materialListBinder;
+    this.viewHolderClass = viewHolderClass;
+    customListData = data;
+  }
```

Comment:
I love that you added support for different collection types! But I still think this implementation could use a bit of polishing.

It feels to me like it would be better to have developers to convert the collection to a `CustomListData` outside of the constructor. This would cut down on duplicate code in your constructors. And if you defined extension functions like `toListData()` on some common collections (array, list, etc) creating a `ListAdapter` would still be really easy for developers.

But speaking of extensions, I'm having trouble seeing the advantage of defining the `CustomListData` class, rather than just using a `List` as bohsen and cketcham suggested. If you go with what I suggested above, it seems like you're basically duplicating the behavior of `List`. Is there a reason you decided to go with this implementation over the other?

Reply:
Good news, MaterialListBinder now integrates a data source into it. Check out the demo for implementation differences.